From patchwork Mon Oct 15 16:41:23 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10642179
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 7D8B0925
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Oct 2018 16:41:32 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 6994929A74
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Oct 2018 16:41:32 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 5A34229A7E; Mon, 15 Oct 2018 16:41:32 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 0333C29A74
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Oct 2018 16:41:32 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726691AbeJPA13 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 15 Oct 2018 20:27:29 -0400
Received: from mail-pf1-f201.google.com ([209.85.210.201]:45557 "EHLO
        mail-pf1-f201.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726525AbeJPA13 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 15 Oct 2018 20:27:29 -0400
Received: by mail-pf1-f201.google.com with SMTP id p89-v6so20650857pfj.12
        for <kvm@vger.kernel.org>; Mon, 15 Oct 2018 09:41:30 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=FLgOApEK059X96s9O/CDyf4l/0n9lBNgZPyg7eBPtI4=;
        b=GhZnzEn2BDFjKyliJKB1JJLZo2TvtaqxjngdZYS5dhhHfiXoFuOwZ3vzU/+x3/10ur
         BH483VsJvYhWZzfhhvd+FDs0wNbdZtZLrabWbLALP5p2Dj4rVoxapyAutLLSCnFz1dlc
         o0Va179m6hcR6P7NbYduQm9EFo6TS39hz8tlyP3yotZPhsMQftOCPpPNMvdfeHKFTACI
         fn6oWoaJJuzifmosOPO5cU+JXMWkAWna8gQ2l3/G9UpRMhDWiBRAP2kDYkEI7GPP2l4G
         UuTHpSEqVXzdQDrSjt9hLrQ/Jq0Kn9Lu9F7Hgo8/pAcEQ7zIqKkqz9mVW2f+PoWYski1
         oVwQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=FLgOApEK059X96s9O/CDyf4l/0n9lBNgZPyg7eBPtI4=;
        b=bbDcRmVuWo0Ez7ctcOVh2A2PH42XK8PQA3DHPmsHNQX+Sz4TL5oU4W9JZeI9g1CTUx
         rkIRc161SXBCkZyczdggvOddxh4+Vy4ZTZOQls1RC0bUPhFXtX5Cp/TBcqTKYwCe9xOO
         Il5RX+ECxjCnaFx4a1bsnqgLPt626uLGCH61xLHA/mTAOhKmvgvJicRQzLIJi1LH3F2Z
         gg12Ehm+NfcntvncS3/sJek8ebLtiMJEAJFCzz6FkfdKv4SlWsdboctWxbEVlmX1h9ao
         5idxnglX4hAvuhKd1FWHfMch/+V/KG5OZ9z46H2FkWdlc3K/jhOJLKWxSaUHXhoSs5cG
         T8qA==
X-Gm-Message-State: ABuFfojr6exxY3zy7deEtNVxV41geTHbi5inCEHPtfYzH4mpfUsVwNpc
        IG5zJ4nOrw6z2X1xReyEvueo0wdNPKu3KUsY50NEscwrxohji8wWmy/JfO27lFxpPvDQRqD6WPb
        7Nkc0+zp7dnEbsSVrOuJZKa1iMT6s1GAHMfdhQhTTkHdvm1awtkKJ2+hxEHU1I8g=
X-Google-Smtp-Source: 
 ACcGV62YOLHcb7kQF1Zgmn47HQ2mCj+EVKUd2DfGcR8XikxwjXn51nh2z18mpnMEuuRNsNYWohS7Q/G/XgfeHg==
X-Received: by 2002:a62:a9b:: with SMTP id 27-v6mr8690249pfk.56.1539621689714;
 Mon, 15 Oct 2018 09:41:29 -0700 (PDT)
Date: Mon, 15 Oct 2018 09:41:23 -0700
Message-Id: <20181015164124.225764-1-jmattson@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.19.1.331.ge82ca0e54c-goog
Subject: [PATCH v2 1/2] kvm: x86: Bounds-check argument to
 x86_exception_has_error_code
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>, Marc Orr <marcorr@google.com>,
        Sean Christopherson <sean.j.christopherson@intel.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

x86_exception_has_error_code should deterministically return false if
its operand is greater than 32.

Fixes: 0447378a4a793 ("kvm: vmx: Nested VM-entry prereqs for event inj.")
Cc: Marc Orr <marcorr@google.com>
Cc: Sean Christopherson <sean.j.christopherson@intel.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
Reviewed-by: Marc Orr <marcorr@google.com>
---
 arch/x86/kvm/x86.h | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/arch/x86/kvm/x86.h b/arch/x86/kvm/x86.h
index 67b9568613f3..152b20d06a48 100644
--- a/arch/x86/kvm/x86.h
+++ b/arch/x86/kvm/x86.h
@@ -116,7 +116,8 @@ static inline bool x86_exception_has_error_code(unsigned int vector)
 			BIT(NP_VECTOR) | BIT(SS_VECTOR) | BIT(GP_VECTOR) |
 			BIT(PF_VECTOR) | BIT(AC_VECTOR);
 
-	return (1U << vector) & exception_has_error_code;
+	WARN_ON_ONCE(vector >= 32);
+	return vector < 32 && (1U << vector) & exception_has_error_code;
 }
 
 static inline bool mmu_is_nested(struct kvm_vcpu *vcpu)

From patchwork Mon Oct 15 16:41:24 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10642181
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 7B4F9925
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Oct 2018 16:41:36 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 69C1729A74
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Oct 2018 16:41:36 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 5DC4629A7E; Mon, 15 Oct 2018 16:41:36 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id D873B29A74
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 15 Oct 2018 16:41:35 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726714AbeJPA1e (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 15 Oct 2018 20:27:34 -0400
Received: from mail-yb1-f201.google.com ([209.85.219.201]:36876 "EHLO
        mail-yb1-f201.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726663AbeJPA1d (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 15 Oct 2018 20:27:33 -0400
Received: by mail-yb1-f201.google.com with SMTP id z1-v6so11110506ybf.4
        for <kvm@vger.kernel.org>; Mon, 15 Oct 2018 09:41:34 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=gH3vOpfOfjsd5JATN5AuGpFIYfOc4r8pj+3P1JHlAwM=;
        b=fn1DfEU08YGQH3Y+5IXM9BWNZYL9RAr6yFtWICxf5GTNH4NfgRz0t8yzLBa7S1dA2S
         VBpu+iEwzRdh77Hmtk12JhIFD1Arm82LLdO5ejS8e00IFsf9zWrZHdGo7fm9CIxxfPil
         fkjKytg9NI7Tx6s74XD0/OhSL4dsW0//0Mb3X+f60mMQ0yWT36wh5jWp77BolyZ1sgVc
         q600eaa0eSQHRy0mSY48XlQgLlclzSLFbps1bsn/g3UBHV5wBafuG4UV7Q2qJNn8oD5G
         eFvtsaQmpGYYimolHlTUfez6t7IDZ9XXqorK2/PeFQHS+TAy/qfx88EwymBNC5hdxhSG
         aBDw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=gH3vOpfOfjsd5JATN5AuGpFIYfOc4r8pj+3P1JHlAwM=;
        b=RDEtzZaEC3zmagO89G4PAlATzR02GXCtmG2DigKH8R0iUYnal6XUEqusLgM3Kzchow
         wU03N2ijYO/JgpP/Rx7Jkr1CrkNd5uyf8t/6bOXLXEtXPpHW5v9sgc0kREtaXQniZQ4m
         +cnm2xU3OHI4FhJZgZacZDhpdL723iyozM7MSNY9+R65TBCcRRWqgzg6p5pOtp8TsNnV
         gwQe7ZxcpE5caXeyG1exDkBJt0i6JUpQt1lGwQ7o6/ubogB7i/tAM6+/7jrMtW3aOoaw
         g2Ut99/nCvxym1rel8APDFjTWKUhyKm5skMPwtesrGL3fwf7NxSAPrkEbwSqfURaH3dr
         Uzyw==
X-Gm-Message-State: ABuFfogJMan/ypoTynHrpMY5+OETL+/cgFAUerVMTvHNjVA9OyQedDKc
        CX+tz6XvFVz14sxIqPqydbCRHdYtvUb+9Qa122EYt9Wx5QO49SrzUkGGIlujHO802F71hxxO8R1
        7hdSeL8M58SKaagj+0OmYEWLmGSGjpb+rp1WkXEKEHWhhvyBYWw/fXILoC8UB3fQ=
X-Google-Smtp-Source: 
 ACcGV61KXOw1VzUKaTzynPqYza2XaRLb9yw1faTx9umhWJmcGG5P+2+a9XiAUMXPFcp/oSnIL3C6tHwDNDCuhA==
X-Received: by 2002:a25:38d1:: with SMTP id
 f200-v6mr9898924yba.59.1539621693748;
 Mon, 15 Oct 2018 09:41:33 -0700 (PDT)
Date: Mon, 15 Oct 2018 09:41:24 -0700
In-Reply-To: <20181015164124.225764-1-jmattson@google.com>
Message-Id: <20181015164124.225764-2-jmattson@google.com>
Mime-Version: 1.0
References: <20181015164124.225764-1-jmattson@google.com>
X-Mailer: git-send-email 2.19.1.331.ge82ca0e54c-goog
Subject: [PATCH v2 2/2] kvm: x86: Disallow invalid exceptions from userspace
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Userspace should not be able to inject an invalid exception into a kvm
guest via KVM_SET_VCPU_EVENTS, nor should it be able to convert a
valid pending exception to an invalid one via KVM_SET_SREGS.

In particular, only certain exceptions deliver an error code in
protected mode, and no exception delivers an error code in
real-address mode.

Reported-by: syzbot <syzkaller@googlegroups.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Marc Orr <marcorr@google.com>
---
 arch/x86/kvm/x86.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 4ab569171ad1..b67a664b1d7b 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -3423,6 +3423,14 @@ static int kvm_vcpu_ioctl_x86_set_vcpu_events(struct kvm_vcpu *vcpu,
 	    vcpu->arch.mp_state == KVM_MP_STATE_INIT_RECEIVED)
 		return -EINVAL;
 
+	if (events->exception.injected) {
+		bool has_error_code = is_protmode(vcpu) &&
+			x86_exception_has_error_code(events->exception.nr);
+
+		if (!!events->exception.has_error_code != has_error_code)
+			return -EINVAL;
+	}
+
 	process_nmi(vcpu);
 	vcpu->arch.exception.injected = false;
 	vcpu->arch.exception.pending = events->exception.injected;
@@ -8170,6 +8178,14 @@ static int kvm_valid_sregs(struct kvm_vcpu *vcpu, struct kvm_sregs *sregs)
 			return -EINVAL;
 	}
 
+	if (vcpu->arch.exception.injected || vcpu->arch.exception.pending) {
+		bool has_error_code = (sregs->cr0 & X86_CR0_PE) &&
+			x86_exception_has_error_code(vcpu->arch.exception.nr);
+
+		if (vcpu->arch.exception.has_error_code != has_error_code)
+			return -EINVAL;
+	}
+
 	return 0;
 }
 
